packages-exp/auth-exp/src/core/strategies/email_and_password.test.ts

```diff
@@ -0,0 +1,293 @@
+      },
+      400
+    );
+    await expect(sendPasswordResetEmail(mockAuth, email)).to.be.rejectedWith(
+      FirebaseError,
+      'Firebase: The email address is badly formatted. (auth/invalid-email).'
+    );
+    expect(mock.calls.length).to.eq(1);
+  });
+
+  context('on iOS', () => {
+    it('should pass action code parameters', async () => {
+      const mock = mockEndpoint(Endpoint.SEND_OOB_CODE, {
+        email
+      });
+      await sendPasswordResetEmail(mockAuth, email, {
+        handleCodeInApp: true,
+        iOS: {
+          bundleId: 'my-bundle',
+          appStoreId: 'my-appstore-id'
+        },
+        url: 'my-url',
+        dynamicLinkDomain: 'fdl-domain'
+      });
+
+      expect(mock.calls[0].request).to.eql({
+        requestType: Operation.PASSWORD_RESET,
+        email,
+        continueUrl: 'my-url',
+        dynamicLinkDomain: 'fdl-domain',
+        canHandleCodeInApp: true,
+        iosBundleId: 'my-bundle',
+        iosAppStoreId: 'my-appstore-id'
+      });
+    });
+  });
+
+  context('on Android', () => {
+    it('should pass action code parameters', async () => {
+      const mock = mockEndpoint(Endpoint.SEND_OOB_CODE, {
+        email
+      });
+      await sendPasswordResetEmail(mockAuth, email, {
+        handleCodeInApp: true,
+        android: {
+          installApp: false,
+          minimumVersion: 'my-version',
+          packageName: 'my-package'
+        },
+        url: 'my-url',
+        dynamicLinkDomain: 'fdl-domain'
+      });
+      expect(mock.calls[0].request).to.eql({
+        requestType: Operation.PASSWORD_RESET,
+        email,
+        continueUrl: 'my-url',
+        dynamicLinkDomain: 'fdl-domain',
+        canHandleCodeInApp: true,
+        androidInstallApp: false,
+        androidMinimumVersionCode: 'my-version',
+        androidPackageName: 'my-package'
+      });
+    });
+  });
+});
+
+describe('confirmPasswordReset', () => {
+  const oobCode = 'oob-code';
+  const newPassword = 'new-password';
+
+  beforeEach(mockFetch.setUp);
+  afterEach(mockFetch.tearDown);
+
+  it('should confirm the password reset and not return the email', async () => {
+    const mock = mockEndpoint(Endpoint.RESET_PASSWORD, {
+      email: 'foo@bar.com'
+    });
+    const response = await confirmPasswordReset(mockAuth, oobCode, newPassword);
+    expect(response).to.be.undefined;
+    expect(mock.calls[0].request).to.eql({
+      oobCode,
+      newPassword
+    });
+  });
+
+  it('should surface errors', async () => {
+    const mock = mockEndpoint(
+      Endpoint.RESET_PASSWORD,
+      {
+        error: {
+          code: 400,
+          message: ServerError.INVALID_OOB_CODE
+        }
+      },
+      400
+    );
+    await expect(
+      confirmPasswordReset(mockAuth, oobCode, newPassword)
+    ).to.be.rejectedWith(
+      FirebaseError,
+      'Firebase: The action code is invalid. This can happen if the code is malformed]: expired]: or has already been used. (auth/invalid-action-code).'
```

Comment:
what's up with `malformed]: expired ]:` ? are those meant to be `,` ?

Reply:
good catch! Apparently we had a bunch of typos in the errors map, probably due to bad regexp on my part